api/v1alpha1/warehouse_types.go

```diff
@@ -117,6 +134,161 @@ type WarehouseSpec struct {
 	FreightCreationCriteria *FreightCreationCriteria `json:"freightCreationCriteria,omitempty" protobuf:"bytes,5,opt,name=freightCreationCriteria"`
 }
 
+var knownKeys = []string{"chart", "git", "image"}
+
+// UnmarshalJSON unmarshals the JSON data into WarehouseSpec, converting the
+// JSON from the Subscriptions field into typed RepoSubscription objects in
+// InternalSubscriptions. Any JSON object with a top-level key other than "git",
+// "image", or "chart" is unpacked into a (generic) Subscription with the key as
+// the ArtifactType.
+func (w *WarehouseSpec) UnmarshalJSON(data []byte) error {
+	type warehouseSpecAlias WarehouseSpec
+	aux := &struct {
+		Subscriptions []apiextensionsv1.JSON `json:"subscriptions"`
+		*warehouseSpecAlias
+	}{
+		warehouseSpecAlias: (*warehouseSpecAlias)(w),
+	}
+
+	if err := json.Unmarshal(data, aux); err != nil {
+		return err
+	}
+
+	// Store the JSON subscriptions
+	w.Subscriptions = aux.Subscriptions
+
+	// Convert JSON Subscriptions to typed RepoSubscription objects
+	if len(aux.Subscriptions) > 0 {
+		w.InternalSubscriptions = make([]RepoSubscription, len(aux.Subscriptions))
+		for i := range aux.Subscriptions {
+			// First, unmarshal as a map to check for keys
+			rawMap := make(map[string]json.RawMessage)
+			if err := json.Unmarshal(aux.Subscriptions[i].Raw, &rawMap); err != nil {
+				return err
+			}
+
+			// Validate that the subscription is an object with exactly one top-level
+			// key
+			if len(rawMap) != 1 {
+				return fmt.Errorf(
+					"subscription at index %d must be an object with exactly one "+
+						"top-level field, but has %d fields",
+					i, len(rawMap),
+				)
+			}
+
+			// Get the single key
+			var key string
+			for k := range rawMap {
+				key = k
+				break // This unnecessary, but it makes it clear what we're doing
+			}
+
+			// Check for known keys (git, image, chart)
+			if slices.Contains(knownKeys, key) {
+				// Known subscription type - unmarshal normally
+				if err := json.Unmarshal(
+					aux.Subscriptions[i].Raw,
+					&w.InternalSubscriptions[i],
+				); err != nil {
+					return err
+				}
+			} else {
+				// Generic subscription - unpack with key as ArtifactType
+				var sub Subscription
+				if err := json.Unmarshal(rawMap[key], &sub); err != nil {
+					return err
+				}
+				sub.SubscriptionType = key
+				w.InternalSubscriptions[i].Subscription = &sub
+			}
+		}
+	}
+
+	return nil
+}
+
+// MarshalJSON marshals the WarehouseSpec to JSON, converting the
+// InternalSubscriptions slice into JSON in the Subscriptions field. For
+// Subscription objects, the Kind field is used as the top-level key.
+func (w *WarehouseSpec) MarshalJSON() ([]byte, error) {
+	type warehouseSpecAlias WarehouseSpec
+
+	// Create a copy to avoid modifying the original
+	spec := *w
+	aux := &struct {
```

Comment:
Super nit: Since we use this inline `struct` twice, should we define it as a private struct here?